Set leader-elect for kube-scheduler to true / Adds new configuration types and rework how we set defaults

dims opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
Thanks to some great sleuthing by ikruglov!

kube-controller-manager defaults --leader-elect to true. We should
do the same for kube-scheduler. kube-scheduler used to have this
set to true, but it got lost during refactoring in:
efb2bb71cd4dede007289634bfe3c51484625ba9

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #59729

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
kube-scheduler has been fixed to use `--leader-elect` option back to true (as it was in previous versions)
```

dims commented:
/assign @timothysc

timothysc reviewed (approved):
/lgtm 
/approve

liggitt commented on cmd/kube-scheduler/app/server.go:
```diff
@@ -150,6 +150,7 @@ func NewOptions() (*Options, error) {
 		return nil, err
 	}
 
+	o.config.LeaderElection.LeaderElect = true
```
seems like defaults should be applied in defaults.go, not here (and options round-tripped through those defaults)

dims replied:
i was trying to match what is done in kube-controller ... will fix https://github.com/kubernetes/kubernetes/blob/master/cmd/cloud-controller-manager/app/options/options.go#L52

liggitt commented:
/hold

liggitt commented:
why is this default different than all the other defaults for the scheduler config, applied in the api defaulting path? the setting here won't get applied when reading from file

dims commented:
@liggitt looks like we need to do both ... cmd/kube-scheduler/app/server.go and pkg/apis/componentconfig/v1alpha1/defaults.go if i don't patch up server.go, then the command line default is still false.

liggitt commented:
> if i don't patch up server.go, then the command line default is still false.

that means the internal config is not being round tripped and defaulted... that seems like it should be fixed as well

dims commented:
@liggitt can you point me to an existing pattern/files where this is done please?

liggitt commented:
> @liggitt can you point me to an existing pattern/files where this is done please?

for kubeproxy:

https://github.com/kubernetes/kubernetes/blob/fdeaa8c67a5a5f63626f6b0399b158dd0ce36dcd/cmd/kube-proxy/app/server.go#L303-L343

for kubelet:

https://github.com/kubernetes/kubernetes/blob/fdeaa8c67a5a5f63626f6b0399b158dd0ce36dcd/cmd/kubelet/app/options/options.go#L255-L268

dims commented:
Thanks @liggitt looks like it was already setup. but there was a mismatch in `bool` vs `*bool` for LeaderElect (in the 2 types.go file) and a hard coding of LeaderElect to false in DefaultLeaderElectionConfiguration.

liggitt commented on pkg/apis/componentconfig/types.go:
```diff
@@ -129,7 +129,7 @@ type LeaderElectionConfiguration struct {
 	// before executing the main loop. Enable this when running replicated
 	// components for high availability.
-	LeaderElect bool
+	LeaderElect *bool
```
internal `bool`, external `*bool` was intentional. Convert_Pointer_bool_To_bool handles the conversion, and external pointer lets you apply defaulting logic if nil.

dims replied:
Ack going back to what i had before

dims committed:
Set leader-elect for kube-scheduler to true

Thanks to some great sleuthing by ikruglov!

kube-controller-manager defaults --leader-elect to true. We should
do the same for kube-scheduler. kube-scheduler used to have this
set to true, but it got lost during refactoring in:
efb2bb71cd4dede007289634bfe3c51484625ba9

dims commented:
/test pull-kubernetes-unit

liggitt commented:
/hold cancel
LGTM

bsalamat commented:
The changes LGTM as well, but I let @timothysc take a look as he is more familiar with the learder election part.

dims commented:
/test pull-kubernetes-unit

dims commented:
@timothysc this is ready now. please take a look. @liggitt helped me whip it into shape

mbohlool commented:
/cc @cheftako

dims commented:
/assign @deads2k 
/assign @mtaufen

deads2k commented:
/lgtm

dims commented:
@mtaufen PTAL, needs your approval for pkg/apis/componentconfig/OWNERS

mtaufen commented on pkg/client/leaderelectionconfig/config.go:
```diff
@@ -33,7 +33,7 @@ const (
 
 func DefaultLeaderElectionConfiguration() componentconfig.LeaderElectionConfiguration {
```
Shouldn't the defaults be produced by running the defaulter on a raw object? This function can easily skew from what the defaulter produces, and the defaulter should be the source of truth.

Additionally, there should not be a concept of an internal object having defaults. Defaults are part of versioned APIs.

I'd expect something like:
```
versioned := &componentconfigv1alpha1.LeaderElectionConfiguration{}
scheme.Default(versioned)
internal := &componentconfig.LeaderElectionConfiguration{}
if err := scheme.Convert(versioned, internal, nil); err != nil {
	return nil, err
}
return internal, nil
```

dims replied:
@mtaufen Done!

mtaufen replied:
You don't have to do it in this PR, but I noticed the pattern of "defaults" divorced from the defaulters is also shared by `k8s.io/kubernetes/cmd/controller-manager/app/options.NewDefaultControllerManagerComponentConfig`, so it would be a good thing to clean that one up too.

dims replied:
Ack @mtaufen i got the hang of it, will give it a shot in another PR

mtaufen replied:
Hmm, actually, now that I look more closely, I guess `LeaderElectionConfiguration` is just a substructure of the `KubeControllerManagerConfiguration`?

It is a bit strange to have a defaulter for a substructure like this. Without the context of *which* component it belongs to, it's premature to set a default. Do we need to merge this PR asap to fix failing tests or can we clean up the defaulting for `KubeControllerManagerConfiguration` here too?

dims replied:
@mtaufen Done! i've cleaned up the defaulting of `KubeControllerManagerConfiguration`

timothysc commented:
/approved

dims commented:
@thockin @luxas @wlan0 - Can one of you please approve for cmd/cloud-controller-manager?
@mikedanese @vishh @mtaufen - Can one of you please approve for pkg/apis/componentconfig?

wlan0 commented:
/approve

dims commented:
@mtaufen @timothysc can you please add this to the 1.10 milestone?

mtaufen commented on cmd/controller-manager/app/options/options.go:
```diff
@@ -90,65 +88,14 @@ func NewGenericControllerManagerOptions(componentConfig componentconfig.KubeCont
 
 // NewDefaultControllerManagerComponentConfig returns default kube-controller manager configuration object.
 func NewDefaultControllerManagerComponentConfig(insecurePort int32) componentconfig.KubeControllerManagerConfiguration {
-	return componentconfig.KubeControllerManagerConfiguration{
```
Did you make sure that all these defaults match the ones applied by the defaulter?

liggitt replied:
Something we've found helpful with similar setups in openshift is having a sanity check unitntest that constructs a config struct, populates it from defaults (or calls this method), then DeepEqual compares it to an expected struct. Setting that unit test up against master, then making sure it passes with no changes against this PR would give me confidence we're not changing anything here. 

That also catches drift or accidental removal of defaults, which is important with config types that are in the process of being moved and are currently shared.

dims replied:
@liggitt there's a `TestAddFlags` in options_test.go in this directory. it has not been touched. it does use `NewCloudControllerManagerOptions`

liggitt replied:
> there's a TestAddFlags in options_test.go in this directory. it has not been touched. it does use NewCloudControllerManagerOptions

that's good, but that test is making sure flags override defaults, not ensuring the defaults match what is expected. if that test also compared against expected defaults before binding/parsing flags, that would cover it

dims replied:
cool. will do that shortly

dims replied:
Done. added a new test without parsing/binding flags. also ran it against master - http://paste.openstack.org/show/718026/ - no errors.

mtaufen commented:
/status approved-for-milestone

mtaufen commented on pkg/apis/componentconfig/v1alpha1/conversion.go:
```diff
@@ -0,0 +1,87 @@
+/*
```
what prevented us from generating these conversions?

dims replied:
@mtaufen i saw some warnings in zz_generated.conversion.go. Are those the things that prevented us from fully auto generating? 

```
zz_generated.conversion.go:	// WARNING: in.LeaderElection requires manual conversion: inconvertible types (*k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1.LeaderElectionConfiguration vs k8s.io/kubernetes/pkg/apis/componentconfig.LeaderElectionConfiguration)
zz_generated.conversion.go:	// WARNING: in.VolumeConfiguration requires manual conversion: inconvertible types (*k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1.VolumeConfiguration vs k8s.io/kubernetes/pkg/apis/componentconfig.VolumeConfiguration)
zz_generated.conversion.go:	// WARNING: in.LeaderElection requires manual conversion: inconvertible types (k8s.io/kubernetes/pkg/apis/componentconfig.LeaderElectionConfiguration vs *k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1.LeaderElectionConfiguration)
zz_generated.conversion.go:	// WARNING: in.VolumeConfiguration requires manual conversion: inconvertible types (k8s.io/kubernetes/pkg/apis/componentconfig.VolumeConfiguration vs *k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1.VolumeConfiguration)
zz_generated.conversion.go:	// WARNING: in.PersistentVolumeRecyclerConfiguration requires manual conversion: inconvertible types (*k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1.PersistentVolumeRecyclerConfiguration vs k8s.io/kubernetes/pkg/apis/componentconfig.PersistentVolumeRecyclerConfiguration)
zz_generated.conversion.go:	// WARNING: in.PersistentVolumeRecyclerConfiguration requires manual conversion: inconvertible types (k8s.io/kubernetes/pkg/apis/componentconfig.PersistentVolumeRecyclerConfiguration vs *k8s.io/kubernetes/pkg/apis/componentconfig/v1alpha1.PersistentVolumeRecyclerConfiguration)
```

mtaufen replied:
Do the substructures need to be pointers in the v1alpha1 type? If not I think you can avoid manual conversions.

dims replied:
Ack. Let me try that @mtaufen

dims commented:
/test all

dims commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu
/test pull-kubernetes-kubemark-e2e-gce
/test pull-kubernetes-node-e2e

dims commented:
/test pull-kubernetes-node-e2e

dims commented:
/test pull-kubernetes-unit

mtaufen commented on cmd/controller-manager/app/options/options.go:
```diff
@@ -90,65 +88,14 @@ func NewGenericControllerManagerOptions(componentConfig componentconfig.KubeCont
-		HorizontalPodAutoscalerUseRESTClients: true,
-	}
+	scheme := runtime.NewScheme()
+	componentconfigv1alpha1.AddToScheme(scheme)
```
Since componentconfig is installed (`k8s.io/kubernetes/pkg/apis/componentconfig/install/install.go`) in legacyscheme (`k8s.io/kubernetes/pkg/api/legacyscheme/scheme.go`), you can probably just use legacyscheme.Scheme instead of creating a new one here.

dims replied:
good one! will try that.

dims replied:
no dice with `legacyscheme.Scheme`, switched back to creating a new scheme

mtaufen replied:
hmm, what failed?

liggitt replied:
new scheme is preferred anyway, rather than polluting a shared s heme with config specific types

mtaufen commented on cmd/controller-manager/app/options/options.go:
```diff
@@ -90,65 +88,14 @@ func NewGenericControllerManagerOptions(componentConfig componentconfig.KubeCont
+	scheme.Default(&versioned)
+	internal := componentconfig.KubeControllerManagerConfiguration{}
+	scheme.Convert(&versioned, &internal, nil)
+	internal.Port = insecurePort
```
this feels out of place here, why not in the defaulter?

dims replied:
this is called from 2 places one from cloud-controller-manager and another from kube-controller-manager with different values for insecurePort. so the default depends on the binary essentially

mtaufen replied:
It might be better not to parameterize it here, and to instead independently override it in code specific to the independent controller-managers.

liggitt replied:
that is probably a sign the config for those two components isn't correct yet. Most of the values don't apply to the cloud controller manager, so I'd probably expect it to get its own config object

dims replied:
i can tweak this in a subsequent PR

mtaufen commented on pkg/apis/componentconfig/v1alpha1/defaults_test.go:
```diff
@@ -41,3 +41,21 @@ func TestSchedulerDefaults(t *testing.T) {
 	}
 }
+
+func TestControllerDefaults(t *testing.T) {
```
maybe `TestControllerDefaultsRoundTrip`?

dims replied:
fixed

mtaufen commented on pkg/client/leaderelectionconfig/config.go:
```diff
@@ -20,25 +20,23 @@ import (
-		ResourceLock:  rl.EndpointsResourceLock,
-	}
+	scheme := runtime.NewScheme()
+	componentconfigv1alpha1.AddToScheme(scheme)
```
same comment here about probably not needing to create a new scheme, because componentconfig should be installed in legacyscheme

dims replied:
actually removed this method since we no longer use it in options.go.

dims commented:
/retest

liggitt commented:
while I'm generally in favor of these changes... this is a large restructure coming in very late. can we do the minimal "default true" change for code freeze and follow up with this PR?

dims commented:
@liggitt ack, going back to my first attempt in this PR :) please see https://github.com/kubernetes/kubernetes/pull/60524

dims commented:
/milestone clear

dims commented:
we merged a small change and pushing this PR off to v1.11

dims commented:
@mtaufen @liggitt let's get this in please? now that master is open for 1.11. thanks!

dims commented:
@liggitt now that master is open and it's early in the cycle, let's please get this in.

liggitt commented on pkg/apis/componentconfig/v1alpha1/types.go:
```diff
@@ -108,9 +108,13 @@ type KubeSchedulerConfiguration struct {
 // LeaderElectionConfiguration defines the configuration of leader election
 // clients for components that can run with leader election enabled.
 type LeaderElectionConfiguration struct {
+	metav1.TypeMeta `json:",inline"`
```
Why does this need type info? It's not a top level type, right?

dims replied:
@liggitt some of the tests fail without it

liggitt replied:
unless this is serialized in standalone way (which I'm not aware of), this should not have type info. tests that fail without that should be fixed.

dims replied:
ack removing it

liggitt commented on pkg/apis/componentconfig/v1alpha1/defaults.go:
```diff
@@ -26,12 +26,181 @@ import (
+		obj.HorizontalPodAutoscalerUseRESTClients = utilpointer.BoolPtr(true)
+	}
+
+	SetDefaults_VolumeConfiguration(&obj.VolumeConfiguration)
```
still outstanding

dims replied:
ack deleting now

dims commented on pkg/apis/componentconfig/types.go:
```diff
@@ -117,15 +117,20 @@ type KubeSchedulerConfiguration struct {
 // LeaderElectionConfiguration defines the configuration of leader election
 // clients for components that can run with leader election enabled.
 type LeaderElectionConfiguration struct {
+	metav1.TypeMeta
```
ack removing it

dims commented:
Ready and all green now @liggitt . thanks!

liggitt commented on pkg/apis/componentconfig/v1alpha1/types.go:
```diff
@@ -146,6 +146,265 @@ type KubeSchedulerLeaderElectionConfiguration struct {
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type KubeControllerManagerConfiguration struct {
```
these should probably end up in `controllermanager.config.k8s.io` under `pkg/controller/apis`, similar to kubelet. can you move them in a follow up?

dims replied:
yes i can

liggitt commented:
thanks
/lgtm

bsalamat commented:
Could you please update the title of this PR and state that it adds new configuration types?
